File: 0337-house-robber-iii/0337-house-robber-iii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rob(self, root: Optional[TreeNode]) -> int:
        def helper(curr):
            if not curr:
                return [0, 0]
            dp = [0] * 2
            
            left = helper(curr.left)
            right = helper(curr.right)

            dp[0] = max(left[0], left[1]) + max(right[0], right[1])
            dp[1] = left[0] + right[0] + curr.val

            return dp

        dp = helper(root)
        return max(dp[0], dp[1])

File: 0337-house-robber-iii/0337-house-robber-iii.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var rob = function(root) {
    const helper = (node) => {
        let dp = [0, 0];
        
        if (node === null) {
            return dp;
        }
        
        let left = helper(node.left);
        let right = helper(node.right);
        
        dp[0] = Math.max(left[0], left[1]) + Math.max(right[0], right[1]);
        dp[1] = node.val + left[0] + right[0];
        
        return dp;
    }
    
    let res = helper(root);
    return Math.max(res[0], res[1]);
};